monty-31/src/dft/mod.rs

```diff
@@ -0,0 +1,296 @@
+        debug_span!("post-transpose", nrows = ncols, ncols = nrows)
+            .in_scope(|| transpose::transpose(&scratch.values, &mut mat.values, nrows, ncols));
+
+        let inv_len = MontyField31::from_canonical_usize(nrows).inverse();
+        debug_span!("scale").in_scope(|| mat.scale(inv_len));
+        mat
+    }
+
+    #[instrument(skip_all, fields(dims = %mat.dimensions(), added_bits))]
+    fn coset_lde_batch(
+        &self,
+        mat: RowMajorMatrix<MontyField31<MP>>,
+        added_bits: usize,
+        shift: MontyField31<MP>,
+    ) -> Self::Evaluations {
+        let nrows = mat.height();
+        let ncols = mat.width();
+        let result_nrows = nrows << added_bits;
+
+        if nrows == 1 {
+            let dupd_rows = core::iter::repeat(mat.values)
+                .take(result_nrows)
+                .flatten()
+                .collect();
+            return RowMajorMatrix::new(dupd_rows, ncols).bit_reverse_rows();
+        }
+
+        let input_size = nrows * ncols;
+        let output_size = result_nrows * ncols;
+
+        // TODO: Use faster bit-reversal algo
+        let mat = mat.bit_reverse_rows().to_row_major_matrix();
+
+        // Allocate space for the output and the intermediate state.
+        // NB: The unsafe version below takes well under 1ms, whereas doing
+        //   vec![MontyField31::zero(); output_size])
+        // takes 100s of ms. Safety is expensive.
+        let (mut output, mut padded) = debug_span!("allocate scratch space").in_scope(|| unsafe {
+            // Safety: These are pretty dodgy, but work because MontyField31 is #[repr(transparent)]
+            let output = transmute::<Vec<u32>, Vec<MontyField31<MP>>>(vec![0u32; output_size]);
+            let padded = transmute::<Vec<u32>, Vec<MontyField31<MP>>>(vec![0u32; output_size]);
+            (output, padded)
+        });
+
+        // `coeffs` will hold the result of the inverse FFT; use the
+        // output storage as scratch space.
+        let coeffs = &mut output[..input_size];
+
+        debug_span!("pre-transpose", nrows, ncols)
+            .in_scope(|| transpose::transpose(&mat.values, coeffs, ncols, nrows));
+
+        // Apply inverse DFT; result is not yet normalised.
+        self.update_inv_twiddles(nrows);
+        let inv_twiddles = self.inv_twiddles.borrow().clone();
+        debug_span!("inverse dft batch", n_dfts = ncols, fft_len = nrows)
+            .in_scope(|| Self::decimation_in_time_dft(coeffs, nrows, &inv_twiddles));
+
+        // At this point the inverse FFT of each column of `mat` appears
+        // as a row in `coeffs`.
+
+        // Normalise inverse DFT and coset shift in one go.
+        // IDEA: consider integrating coset shift into twiddles; the current timing
+        // suggests this is not worth the effort.
+        let inv_len = MontyField31::from_canonical_usize(nrows).inverse();
+        coset_shift_and_scale_rows(&mut padded, result_nrows, coeffs, nrows, shift, inv_len);
```

Comment:
This could potentially be made cheaper if we optimized our choice of shift. (Maybe make an issue with this an the above idea in the comment above let inv_len ... about integrating the coset shifts into the twiddles if the cost of this is noticeable?)